src/execution_plan/execution_plan_reduce_to_apply.c

```diff
@@ -0,0 +1,164 @@
+		arguments = ExecutionPlan_BuildArgumentModifiesArray(bound_vars);
+	}
+	OpBase *arg = NewArgumentOp(plan, arguments);
+
+	/* In case of Apply operations, we need to append, the new argument branch and
+	 * set it as the first child. */
+	if(op->type & (OPType_APPLY_MULTIPLEXER | OPType_SEMI_APPLY)) {
+		ExecutionPlan_AddOp(op, arg);
+		_OpBaseSwapChildren(op, 0, op->childCount - 1);
+	} else {
+		// -1 & everything is true;
+		op = ExecutionPlan_LocateFirstOp(op, -1);
+		ExecutionPlan_AddOp(op, arg);
+	}
+}
+
+/* This method reduces a filter tree into an OpBase. The method perfrom post-order traversal over the
+ * filter tree, and checks if if the current subtree rooted at the visited node contains path filter or not,
+ * and either reduces the root or continue traversal and reduction.
+ * The three possible operations could be:
+ * 1. OpFilter - In the case the subtree originated from the root does not contains any path filters.
+ * 2. OpSemiApply - In case of the current root is an expression which contains a path filter.
+ * 3. ApplyMultiplexer - In case the current root is an operator (OR or AND) and at least one of its children
+ * has been reduced to OpSemiApply or ApplyMultiplexer. */
+static OpBase *_ExecutionPlan_FilterTreeToOpBaseReduction(ExecutionPlan *plan,
+														  FT_FilterNode *filter_root) {
+	FT_FilterNode *node;
+	// Case of an expression which contains path filter.
+	if(filter_root->t == FT_N_EXP && FilterTree_containsFunc(filter_root, "path_filter", &node)) {
+		AR_ExpNode *expression = filter_root->exp.exp;
+		bool anti = false;
+		if(strcasecmp(expression->op.func_name, "not") == 0) {
+			anti = true;
+			expression = expression->op.children[0];
+		}
+		// Build new Semi Apply op.
+		OpBase *op_semi_apply = NewSemiApplyOp(plan, anti);
+		const cypher_astnode_t *path = expression->op.children[0]->operand.constant.ptrval;
+		// Add a match branch as a Semi Apply op child.
+		ExecutionPlan_AddOp(op_semi_apply, _buildMatchBranch(plan, path));
+		return op_semi_apply;
+	}
+	// Case of an operator (Or or And) which its subtree contains path filter
+	if(filter_root->t == FT_N_COND && FilterTree_containsFunc(filter_root, "path_filter", &node)) {
+		// Create the relevant LHS branch and set a bounded branch for it.
+		OpBase *lhs = _ExecutionPlan_FilterTreeToOpBaseReduction(plan, filter_root->cond.left);
+		_CreateBoundBranch(lhs, plan);
+		// Create the relevant RHS branch and set a bounded branch for it.
+		OpBase *rhs = _ExecutionPlan_FilterTreeToOpBaseReduction(plan, filter_root->cond.right);
+		_CreateBoundBranch(rhs, plan);
+		// Create multiplexer op and set the branches as its children.
+		OpBase *apply_multiplexer = NewApplyMultiplexerOp(plan, filter_root->cond.op);
+		ExecutionPlan_AddOp(apply_multiplexer, lhs);
+		ExecutionPlan_AddOp(apply_multiplexer, rhs);
+		return apply_multiplexer;
+	}
+	return NewFilterOp(plan, FilterTree_Clone(filter_root));
```

Comment:
Do we actually need to clone the filter tree, given that the original will be freed at 163? Why not just assign the pointer here and NULL it in the original?

Reply:
nice
done